packages/inter-protocol/src/provisionPoolKit.js

```diff
@@ -327,62 +341,96 @@ export const prepareProvisionPoolKit = (
+         * @param {ERef<Brand>} brand
          */
-        start({ metrics } = {}) {
+        watchCurrentAmount(exchangePurse, brand) {
           const {
             state: { brandToPSM, poolBrand },
             facets: { helper },
           } = this;
 
-          // Must match. poolBrand is from durable state and the param is from
-          // the contract, so it technically can change between incarnations.
-          // That would be a severe bug.
-          AmountMath.coerce(poolBrand, params.getPerAccountInitialAmount());
+          void observeNotifier(E(exchangePurse).getCurrentAmountNotifier(), {
+            updateState: async amount => {
+              trace('provisionPool balance update', amount);
+              if (AmountMath.isEmpty(amount) || amount.brand === poolBrand) {
+                return;
+              }
+              if (!brandToPSM.has(brand)) {
+                console.error('funds arrived but no PSM instance', brand);
+                return;
+              }
+              const instance = brandToPSM.get(brand);
+              const payment = E(exchangePurse).withdraw(amount);
+              await helper
+                .swap(payment, amount, instance)
+                .catch(async reason => {
+                  console.error(X`swap failed: ${reason}`);
+                  const resolvedPayment = await payment;
+                  return E(exchangePurse).deposit(resolvedPayment);
+                });
+            },
+            fail: reason => {
+              if (isUpgradeDisconnection(reason)) {
+                void helper.watchCurrentAmount(exchangePurse, brand);
+              } else {
+                console.error(reason);
+              }
+            },
+          });
+        },
+        watchAssetSubscription() {
+          const { facets } = this;
+          const { helper } = facets;
 
-          void observeIteration(
+          return observeIteration(
             subscribeEach(E(poolBank).getAssetSubscription()),
             {
               updateState: async desc => {
-                console.log('provisionPool notified of new asset', desc.brand);
-                await zcf.saveIssuer(desc.issuer, desc.issuerName);
+                await null;
+                const issuer = zcf.getTerms().issuers[desc.issuerName];
+                if (issuer !== desc.issuer) {
+                  if (desc.issuerName.match(FIRST_CAP_ASCII)) {
+                    trace(`Saved Issuer ${desc.issuerName}`);
+                    await zcf.saveIssuer(desc.issuer, desc.issuerName);
+
+                    // see https://github.com/Agoric/agoric-sdk/issues/8238
+                  } else if (desc.issuerName.match(FIRST_LOWER_ASCII)) {
+                    await saveIssuerWithLegalKeyword(desc, zcf);
+                  } else {
+                    console.error(
+                      `unable to save issuer with illegal keyword: ${desc.issuerName}`,
+                    );
+                  }
```

Comment:
both branch save with a legal keyword. This abstraction makes it look like this branch does something different. I think it would be more clear if you made a valid `desc` and then saved it consistently.

```suggestion
                  // let this handle the conditions and log about whether repair was necessary  
                  const goodDesc = repair(desc);
                  if (goodDesc) {
                    await zcf.saveIssuer(desc.issuer, desc.issuerName);
                  } else {
                    console.error(
                      `unable to save issuer with illegal keyword: ${desc.issuerName}`,
                    );
                  }
```

Reply:
Yes, that's better.